src/webgpu/api/operation/resource_init/buffer.spec.ts

```diff
@@ -5,7 +5,276 @@ export const description = `
+    for (let i = 0; i < copySize; ++i) {
+      expectedData[appliedOffset + i] = writeData[i] = i + 1;
+    }
+    t.queue.writeBuffer(buffer, appliedOffset, writeData, 0);
+
+    t.expectContents(buffer, expectedData);
+  });
+
+g.test('map_read_whole_buffer').fn(async t => {
+  const bufferSize = 32;
+
+  const buffer = t.device.createBuffer({
+    size: bufferSize,
+    usage: GPUBufferUsage.COPY_DST | GPUBufferUsage.MAP_READ,
+  });
+
+  await buffer.mapAsync(GPUMapMode.READ);
+  const readData = new Uint8Array(buffer.getMappedRange());
+  for (let i = 0; i < readData.length; ++i) {
+    t.expect(readData[i] === 0);
+  }
+  buffer.unmap();
+});
+
+g.test('map_read_partial_buffer')
+  .paramsSubcasesOnly(u => u.combine('offset', [0, 8, -16]))
+  .fn(async t => {
+    const { offset } = t.params;
+    const bufferSize = 32;
+    const appliedOffset = offset >= 0 ? offset : bufferSize + offset;
+
+    const buffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.COPY_DST | GPUBufferUsage.MAP_READ,
+    });
+
+    {
+      const mapSize = 16;
+      await buffer.mapAsync(GPUMapMode.READ, appliedOffset, mapSize);
+      const readData = new Uint8Array(buffer.getMappedRange(appliedOffset, mapSize));
+      for (let i = 0; i < mapSize; ++i) {
+        t.expect(readData[i] === 0);
+      }
+      buffer.unmap();
+    }
+
+    {
+      await buffer.mapAsync(GPUMapMode.READ);
+      const readData = new Uint8Array(buffer.getMappedRange());
+      for (let i = 0; i < bufferSize; ++i) {
+        t.expect(readData[i] === 0);
+      }
+      buffer.unmap();
+    }
+  });
+
+g.test('map_write_whole_buffer').fn(async t => {
+  const bufferSize = 32;
+  const buffer = t.device.createBuffer({
+    size: bufferSize,
+    usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.MAP_WRITE,
+  });
+
+  await buffer.mapAsync(GPUMapMode.WRITE);
+  const writeData = new Uint8Array(buffer.getMappedRange());
+  for (let i = 0; i < bufferSize; ++i) {
+    t.expect(writeData[i] === 0);
+  }
+  buffer.unmap();
+});
+
+g.test('map_write_partial_buffer')
+  .paramsSubcasesOnly(u => u.combine('offset', [0, 8, -16]))
+  .fn(async t => {
+    const { offset } = t.params;
+    const bufferSize = 32;
+    const appliedOffset = offset >= 0 ? offset : bufferSize + offset;
+
+    const buffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.MAP_WRITE,
+    });
+
+    {
+      const mappedSize = 12;
+      await buffer.mapAsync(GPUMapMode.WRITE, appliedOffset, mappedSize);
+      const writeData = new Uint8Array(buffer.getMappedRange(appliedOffset, mappedSize));
+      for (let i = 0; i < mappedSize; ++i) {
+        t.expect(writeData[i] === 0);
+      }
+      buffer.unmap();
+    }
+
+    {
+      const expectedData = new Uint8Array(bufferSize);
+      t.expectContents(buffer, expectedData);
+    }
+  });
+
+g.test('mapped_at_creation_map_readable_whole_buffer').fn(async t => {
```

Comment:
I think the mapped_at_creation tests could be relatively easy to parametrize instead of duplicating code.